### PR TITLE
fix: yeoman-environment version update + code refactor + tests

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

**Comment by slavik-lvovsky** [comment]
fixes this bug: https://sapjira.wdf.sap.corp/browse/DEVXBUGS-7878

File: backend/src/contributors.ts [hunk 1]
```diff
@@ -2,56 +2,39 @@ import * as vscode from 'vscode';
...
-		return snippet;
-	}
-
-    public static add(extensionId: string, api: any) {
-        Contributors.apiMap.set(extensionId, api);
+    public static async getSnippet(uiOptions: any) {
+        const contributorId = _.get(uiOptions, "contributorId");
+        const snippetName = _.get(uiOptions, "snippetName");
+        if (contributorId && snippetName) {
+            const api = await Contributors.apiMap.get(contributorId);
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:11]: feel to me like there is no await for the activate() method.

**Review by tomer-epstein — CHANGES_REQUESTED** [review]

File: backend/src/contributors.ts (hunk 1 above)
Comment (tomer-epstein) [reply backend/src/contributors.ts:11]: I would like to keep a getApi method.

File: backend/src/panels/CodeSnippetPanel.ts [hunk 2]
```diff
@@ -25,25 +25,27 @@ export class CodeSnippetPanel extends AbstractWebviewPanel {
...
-			this.logger, 
-			{messages: this.messages, snippet: this.snippet});
-		this.codeSnippet.registerCustomQuestionEventHandler("file-browser", "getFilePath", this.showOpenFileDialog.bind(this));
-		this.codeSnippet.registerCustomQuestionEventHandler("folder-browser", "getPath", this.showOpenFolderDialog.bind(this));
-
-		this.initWebviewPanel();
+		Contributors.getSnippet(_.get(uiOptions, "snippet", uiOptions)).then(snippet => {
+			if (_.isNil(snippet)) {
+				this.webViewPanel.dispose();
+				return vscode.window.showErrorMessage("Can not find snippet.");
```
Comment (tomer-epstein) [inline backend/src/panels/CodeSnippetPanel.ts:31]: I consider remove the vscode notification. and only only log it.

File: backend/src/panels/CodeSnippetPanel.ts [hunk 3]
```diff
@@ -25,25 +25,27 @@ export class CodeSnippetPanel extends AbstractWebviewPanel {
...
-			this.outputChannel, 
-			this.logger, 
-			{messages: this.messages, snippet: this.snippet});
-		this.codeSnippet.registerCustomQuestionEventHandler("file-browser", "getFilePath", this.showOpenFileDialog.bind(this));
-		this.codeSnippet.registerCustomQuestionEventHandler("folder-browser", "getPath", this.showOpenFolderDialog.bind(this));
-
-		this.initWebviewPanel();
+		Contributors.getSnippet(_.get(uiOptions, "snippet", uiOptions)).then(snippet => {
+			if (_.isNil(snippet)) {
+				this.webViewPanel.dispose();
```
Comment (tomer-epstein) [inline backend/src/panels/CodeSnippetPanel.ts:30]: Nice 💯

File: backend/src/contributors.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts:11]: extension.activate() returns Thenable, so it is OK to use await

Comment (slavik-lvovsky) [reply backend/src/contributors.ts:11]: I have added a new test with activate method

File: backend/src/panels/CodeSnippetPanel.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply backend/src/panels/CodeSnippetPanel.ts:31]: We talked to leave it, didn't we ?

File: backend/src/contributors.ts (hunk 1 above)
Comment (tomer-epstein) [reply backend/src/contributors.ts:11]: const apiPromise = Contributors.apiMap.get(contributorId);
            const api = await

**Review by tomer-epstein — CHANGES_REQUESTED** [review]

File: backend/src/contributors.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts]: Done

File: backend/src/contributors.ts [hunk 4]
```diff
@@ -2,56 +2,44 @@ import * as vscode from 'vscode';
...
-                    const extensionName: string =  _.get(currentPackageJSON, "name");
-                    const extensionPublisher: string =  _.get(currentPackageJSON, "publisher");
-                    const extensionId: string = extensionPublisher + "." + extensionName;
-                    const api = await Contributors.getApi(extension, extensionId);
-                    Contributors.add(extensionId, api);
-                }
-            }
-        }
+    private static getApiPromise(extension: vscode.Extension<any>) {
+        return (extension.isActive ? extension.exports : extension.activate());
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:34]: extension.exports is not a promise, but we can keep the name.

**Comment by tomer-epstein** [comment]
as we just discuss. lets activate (if not already active) only when trying to get the snippet, and then update the apiMap.

File: backend/src/contributors.ts [hunk 5]
```diff
@@ -2,43 +2,38 @@ import * as vscode from 'vscode';
...
-    public static init() {
-        _.forEach(vscode.extensions.all, (extension: vscode.Extension<any>) => {
+    private static getContributorExtension(contributorId: string) {
+        return _.find(vscode.extensions.all, (extension: vscode.Extension<any>) => {
             const extensionDependencies: string[] = _.get(extension, "packageJSON.extensionDependencies");
             if (_.includes(extensionDependencies, "saposs.code-snippet")) {
-                Contributors.add(extension);
+                const extensionName: string = _.get(extension, "packageJSON.name");
+                const extensionPublisher: string = _.get(extension, "packageJSON.publisher");
+                const extensionId = `${extensionPublisher}.${extensionName}`;
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:33]: private static getExtensionId(extension: vscode.Extension<any>)

File: backend/src/panels/CodeSnippetPanel.ts [hunk 6]
```diff
@@ -25,22 +25,22 @@ export class CodeSnippetPanel extends AbstractWebviewPanel {
 	public setWebviewPanel(webViewPanel: vscode.WebviewPanel, uiOptions?: any) {
 		super.setWebviewPanel(webViewPanel, uiOptions);
 
-		Contributors.getSnippet(_.get(uiOptions, "snippet", uiOptions)).then(snippet => {
+		const contributerInfo = _.get(uiOptions, "contributerInfo", uiOptions);
```
Comment (tomer-epstein) [inline backend/src/panels/CodeSnippetPanel.ts:28]: This will break all code-snippet contributors.

File: backend/src/panels/CodeSnippetPanel.ts [hunk 7]
```diff
@@ -25,22 +25,22 @@ export class CodeSnippetPanel extends AbstractWebviewPanel {
...
+			this.messages = _.assign({}, backendMessages, snippet.getMessages());
 			const rpc = new RpcExtension(this.webViewPanel.webview);
 			this.outputChannel = new OutputChannelLog(this.messages.channelName);
 			const vscodeEvents: AppEvents = new VSCodeEvents(rpc, this.webViewPanel);
 			this.codeSnippet = new CodeSnippet(rpc,
 				vscodeEvents,
 				this.outputChannel,
 				this.logger,
-				{ messages: this.messages, snippet: this.snippet });
+				{ messages: this.messages, snippet, contributerInfo});
```
Comment (tomer-epstein) [inline backend/src/panels/CodeSnippetPanel.ts:43]: ?

File: backend/src/contributors.ts [hunk 8]
```diff
@@ -2,43 +2,38 @@ import * as vscode from 'vscode';
...
+                const errorMessage = _.get(error, "stack", _.get(error, "message", error));
+                console.error(errorMessage);
+                // TODO: Add Logger.error
+            }
         }
     }
 
     private static getApiPromise(extension: vscode.Extension<any>) {
-        return (extension.isActive ? extension.exports : extension.activate());
+        return (extension.isActive ? Promise.resolve(extension.exports) : extension.activate());
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:24]: no need to wrap extension.exports with Promise.

File: backend/src/contributors.ts [hunk 9]
```diff
@@ -2,43 +2,38 @@ import * as vscode from 'vscode';
...
-            Contributors.apiMap.set(extensionId, apiPromise);
-        } catch (error) {
-            const errorMessage = _.get(error, "stack", _.get(error, "message", error));
-            console.error(errorMessage);
-            // TODO: Add Logger.error
+    public static async getSnippet(contributerInfo: any) {
+        const contributorId = _.get(contributerInfo, "contributorId");
+        const snippetName = _.get(contributerInfo, "snippetName");
+        const extension = Contributors.getContributorExtension(contributorId);
+        if (extension) {
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:9]: what if extension not found?

**Review by tomer-epstein — CHANGES_REQUESTED** [review]

File: backend/src/contributors.ts (hunk 9 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts:9]: in this case tab is closed and we show error message

File: backend/src/contributors.ts [hunk 10]
```diff
@@ -2,56 +2,39 @@ import * as vscode from 'vscode';
...
-    public static add(extensionId: string, api: any) {
-        Contributors.apiMap.set(extensionId, api);
-    }
-
-    private static async getApi(extension: vscode.Extension<any>, extensionId: string) {
-        let api: any;
-        if (!extension.isActive) {
+    public static async getSnippet(contributerInfo: any) {
+        const contributorId = _.get(contributerInfo, "contributorId");
+        const snippetName = _.get(contributerInfo, "snippetName");
```
Comment (tomer-epstein) [inline backend/src/contributors.ts:7]: move into the try block

File: backend/src/contributors.ts (hunk 5 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts]: done

File: backend/src/contributors.ts (hunk 8 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts]: done

File: backend/src/contributors.ts (hunk 10 above)
Comment (slavik-lvovsky) [reply backend/src/contributors.ts]: done

**Review by tomer-epstein — DISMISSED** [review]

**Review by tomer-epstein — APPROVED** [review]

**Pull request merged**